### PR TITLE
Improve TestMultiRenderTarget2 test Demo

### PR DESCRIPTION
**Pull request by JohnLKkk**
Improve TestMultiRenderTarget2 test Demo

### CODE REVIEW

**Comment by stephengold** [comment]
Thanks for creating this PR.

It needs some additional documentation. The class-level javadoc should say what the test does and how to tell whether the test is successful ("Displays 4 pictures, each showing a different render of a hovertank."). The pictures don't have descriptive names, so comments should explain how each one differs from the others. Also some explanation, in "MRTLighting.j3md", of what the resulting material does.

I also have some line-by-line review, which will appear separately.

File: jme3-examples/src/main/java/jme3test/post/TestMultiRenderTarget2.java [hunk 1]
```diff
@@ -45,110 +45,139 @@
...
 import com.jme3.texture.Texture2D;
-import com.jme3.texture.FrameBuffer.FrameBufferTarget;
 import com.jme3.ui.Picture;
+import com.jme3.util.mikktspace.MikktspaceTangentGenerator;
 
-public class TestMultiRenderTarget extends SimpleApplication implements SceneProcessor {
+/**
+ * Note that this example is only used to demonstrate the simple use of multi-target rendering. It is different from optimized deferred rendering.<br/>
+ * @author JhonKkk
+ * @date 2021年11月7日10点18分
```
Comment (stephengold) [inline jme3-examples/src/main/java/jme3test/post/TestMultiRenderTarget2.java:59]: "date" is an unknown javadoc tag. Remove it.

File: jme3-examples/src/main/java/jme3test/post/TestMultiRenderTarget2.java [hunk 2]
```diff
@@ -45,110 +45,139 @@
...
 import com.jme3.texture.FrameBuffer;
 import com.jme3.texture.Image.Format;
 import com.jme3.texture.Texture2D;
-import com.jme3.texture.FrameBuffer.FrameBufferTarget;
 import com.jme3.ui.Picture;
+import com.jme3.util.mikktspace.MikktspaceTangentGenerator;
 
-public class TestMultiRenderTarget extends SimpleApplication implements SceneProcessor {
+/**
+ * Note that this example is only used to demonstrate the simple use of multi-target rendering. It is different from optimized deferred rendering.<br/>
```
Comment (stephengold) [inline jme3-examples/src/main/java/jme3test/post/TestMultiRenderTarget2.java:57]: The "br/" tag is not allowed in javadoc.

File: jme3-examples/src/main/java/jme3test/post/TestMultiRenderTarget2.java [hunk 3]
```diff
@@ -45,110 +45,139 @@
...
+ * @author JhonKkk
+ * @date 2021年11月7日10点18分
+ */
+public class TestMultiRenderTarget2 extends SimpleApplication implements SceneProcessor {
 
-    private FrameBuffer fb;
-    private Texture2D diffuseData;
+    private FrameBuffer gBuffFrameBuffer;
+    private Texture2D diffuseData, normalData, specularData, depthData;
```
Comment (stephengold) [inline jme3-examples/src/main/java/jme3test/post/TestMultiRenderTarget2.java:64]: Put each declaration on a separate line.

File: jme3-examples/src/main/java/jme3test/post/TestMultiRenderTarget2.java [hunk 4]
```diff
@@ -45,110 +45,139 @@
...
     private Picture display1, display2, display3, display4;
```
Comment (stephengold) [inline jme3-examples/src/main/java/jme3test/post/TestMultiRenderTarget2.java:65]: Put each declaration on a separate line, with a descriptive comment.

File: jme3-examples/src/main/java/jme3test/post/TestMultiRenderTarget2.java [hunk 5]
```diff
@@ -45,110 +45,139 @@
...
         cam.setRotation(new Quaternion(-0.05143692f, 0.9483723f, -0.21131563f, -0.230846f));
 
-        Node tank = (Node) assetManager.loadModel("Models/HoverTank/Tank2.mesh.xml");
-        
-        //tankMesh.getMaterial().setColor("Specular", ColorRGBA.Black);
-        rootNode.attachChild(tank);
-
         display1 = new Picture("Picture");
-        display1.move(0, 0, -1); // make it appear behind stats view
+        display1.move(0, 0, -1);
```
Comment (stephengold) [inline jme3-examples/src/main/java/jme3test/post/TestMultiRenderTarget2.java:88]: This statement is unnecessary. Please remove it.

File: jme3-examples/src/main/java/jme3test/post/TestMultiRenderTarget2.java [hunk 6]
```diff
@@ -45,110 +45,139 @@
...
+                ColorRGBA.Green,
+                ColorRGBA.Magenta,
+                ColorRGBA.Orange,
+                ColorRGBA.Pink,
+                ColorRGBA.Red,
+                ColorRGBA.Yellow
         };
 
-        pls = new PointLight[3];
+        // lights
```
Comment (stephengold) [inline jme3-examples/src/main/java/jme3test/post/TestMultiRenderTarget2.java:107]: This section adds 20 lights to the scene. Why?

File: jme3-examples/src/main/java/jme3test/post/TestMultiRenderTarget2.java [hunk 7]
```diff
@@ -45,110 +45,139 @@
...
             display.addLight(pl);
             pls[i] = pl;
         }
+
+        // test model
+        Node scene = new Node("TestScene");
+        Geometry model = (Geometry) ((Node)assetManager.loadModel("Models/HoverTank/Tank2.mesh.xml")).getChild(0);
+        MikktspaceTangentGenerator.generate(model);
+        // GBuffer Pass
+        Material tankMat = assetManager.loadMaterial("jme3test/post/tank.j3m");
```
Comment (stephengold) [inline jme3-examples/src/main/java/jme3test/post/TestMultiRenderTarget2.java:125]: What does "tankMat" do?

File: jme3-examples/src/main/java/jme3test/post/TestMultiRenderTarget2.java [hunk 8]
```diff
@@ -45,110 +45,139 @@
...
+        initGBufferInfos();
+
+        rootNode.attachChild(scene);
+        guiViewPort.setClearFlags(true, true, true);
+        guiNode.attachChild(display1);
+        guiNode.attachChild(display2);
+        guiNode.attachChild(display3);
+        guiNode.attachChild(display4);
+        guiNode.attachChild(display);
+        guiNode.updateGeometricState();
```
Comment (stephengold) [inline jme3-examples/src/main/java/jme3test/post/TestMultiRenderTarget2.java:138]: Explicit updating is not needed here. Delete it.

File: jme3-examples/src/main/resources/jme3test/post/MRTLighting.j3md [hunk 9]
```diff
@@ -0,0 +1,370 @@
+MaterialDef MRTLighting {
```
Comment (stephengold) [inline jme3-examples/src/main/resources/jme3test/post/MRTLighting.j3md:1]: Add explanatory comments here.

File: jme3-examples/src/main/java/jme3test/post/TestMultiRenderTarget2.java [hunk 10]
```diff
@@ -45,110 +45,139 @@
...
+        // test model
+        Node scene = new Node("TestScene");
+        Geometry model = (Geometry) ((Node)assetManager.loadModel("Models/HoverTank/Tank2.mesh.xml")).getChild(0);
+        MikktspaceTangentGenerator.generate(model);
+        // GBuffer Pass
+        Material tankMat = assetManager.loadMaterial("jme3test/post/tank.j3m");
+        model.setMaterial(tankMat);
+        scene.attachChild(model);
+        // Create FrameBuffer for multi-target rendering
+        initGBufferInfos();
```
Comment (stephengold) [inline jme3-examples/src/main/java/jme3test/post/TestMultiRenderTarget2.java:129]: What does the "G" in `initGBufferInfos()` stand for?

**Review by stephengold — CHANGES_REQUESTED** [review]

**Comment by stephengold** [comment]
Also, when I run the new test, I get warnings. If possible, please modify the test to avoid these:
```text
Nov 17, 2021 11:16:02 AM com.jme3.material.plugins.J3MLoader readDefine
WARNING: In technique 'LightPass':
Define 'ATTENUATION' mapped to non-existent material parameter 'Attenuation', ignoring.
Nov 17, 2021 11:16:02 AM com.jme3.material.plugins.J3MLoader readDefine
WARNING: In technique 'LightPass':
Define 'HQ_ATTENUATION' mapped to non-existent material parameter 'HighQuality', ignoring.
```

**Comment by stephengold** [comment]
@JohnLKkk are you still available to work on this?

**Comment by JohnLKkk** [comment]
Sorry, I'm a bit busy this time, I will check the warning message here.

**Comment by JohnLKkk** [comment]
> Sorry, I'm a bit busy this time, I will check the warning message here.

**Pull request closed without merging**